File: Difficulty: Medium/Indexes of Subarray Sum/indexes-of-subarray-sum.java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;

class Main{
	static BufferedReader br;
    static PrintWriter ot;
    public static void main(String[] args) throws IOException{
        
        br = new BufferedReader(new InputStreamReader(System.in));
        ot = new PrintWriter(System.out);

        int t = Integer.parseInt(br.readLine());

        while(t-->0){
            
            String s[] = br.readLine().trim().split(" ");
            
            int n = Integer.parseInt(s[0]);
            int k = Integer.parseInt(s[1]);
            int a[] = new int[n];
            s = br.readLine().trim().split(" ");
            for(int i = 0; i < n; i++)
                a[i] = Integer.parseInt(s[i]);
            Solution obj = new Solution();
            ArrayList<Integer> res = obj.subarraySum(a, n, k);
            for(int ii = 0;ii<res.size();ii++)
                ot.print(res.get(ii) + " ");
            ot.println();
        }
        ot.close();
    }

}
// } Driver Code Ends


class Solution
{
    //Function to find a continuous sub-array which adds up to a given number.
    static ArrayList<Integer> subarraySum(int[] arr, int n, int s) 
    {
        // Your code here
        ArrayList<Integer> ans = new ArrayList<>();
        
        if(n == 1){
            if(arr[0] == s){
                ans.add(n);
                ans.add(n);
            }
            else{
                ans.add(-1);
            }
            return ans;
        }
        
        
        //O(n2) --> tle
        
        // for(int i=0;i<n;i++){
        //     int sum = 0;
        //     for(int j=i;j<n;j++){
        //         sum += arr[j];
        //         if(sum == s){
        //             ans.add(i+1);
        //             ans.add(j+1);
        //             return ans;
        //         }
                
        //     }
        // }
        
        
        //using sliding window
        int sum = 0;
        int j = 0;
        for(int i=0;i<n;i++){
            sum += arr[i];
            
            while(sum>s && j<i){
                sum -= arr[j];
                j++;
            }
            
            if(sum == s){
                ans.add(j+1);
                ans.add(i+1);
                return ans;
            }
        }
        
        ans.add(-1);
        
        return ans;
        
    }
}

File: Difficulty: Medium/Indexes of Subarray Sum/indexes-of-subarray-sum.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    vector<int> subarraySum(vector<int> &arr, int target) {
        // code here
        //brute force
        vector<int> ans;
        int n = arr.size();
        
        // if(n == 1){
        //     if(arr[0] == target){
        //         ans.push_back(1);
        //         ans.push_back(1);
        //     }
        //     else{
        //         ans.push_back(-1);
        //     }
        //     return ans;
        // }
        
        
        
        //brute force = O(n2)
        // for(int i=0;i<n;i++){
        //     int sum = 0;
        //     for(int j=i;j<n;j++){
        //         sum += arr[j];
        //         if(sum == target){
        //             ans.push_back(i+1);
        //             ans.push_back(j+1);
        //             return ans;
        //         }
                
        //     }
        // }
        
        //optimised 
        //sliding window
        
        int left = 0;
        int sum = 0;
        
        for(int i=0;i<n;i++){
            sum+=arr[i];
            
            while(sum > target && left<i){
                sum-=arr[left];
                left++;
            }
            
            if(sum == target){
                ans.push_back(left+1);
                ans.push_back(i+1);
                
                return ans;
            }
        }
        
       ans.push_back(-1);
       return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        vector<int> arr;
        int d;
        string input;

        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }

        cin >> d;
        cin.ignore();

        Solution ob;
        vector<int> result = ob.subarraySum(arr, d);
        for (int i : result) {
            cout << i << " ";
        }
        cout << "\n~\n";
    }
    return 0;
}

// } Driver Code Ends